src/core/lbgpu_cuda.cu

```diff
@@ -1020,6 +1061,67 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
 }
 
 
+/* bounce_back function used in bounce_back_boundaries
+** called 18times, once for each velocity
+*/
+__device__ __inline__ void bounce_back(LB_nodes_gpu n_curr, LB_rho_v_gpu *d_v, float *boundary_force, float* torque, unsigned int index, float weight,  int *c, float *v, float* omega, int population, int inverse, short moving_flag, float *to_center, unsigned int x, unsigned int y, unsigned int z){
+  
+  size_t to_index, to_index_x, to_index_y, to_index_z;
+  float r[3];
+  float loc_v[3];
+  float f[3];
+  float shift, pop_to_bounce_back;
+  float vc;
+  
+#ifndef SHANCHEN
+  //+ 4.5f*uc*uc + u_sq_neg
+  pop_to_bounce_back =  n_curr.vd[population*para.number_of_nodes + index ];
+  to_index_x = (x+c[0]+para.dim_x)%para.dim_x;
+  to_index_y = (y+c[1]+para.dim_y)%para.dim_y;
+  to_index_z = (z+c[2]+para.dim_z)%para.dim_z;
+  to_index = to_index_x + para.dim_x*to_index_y + para.dim_x*para.dim_y*to_index_z;
+  if ( n_curr.boundary[to_index] == 0)
+  {
+    r[0] = to_center[0] - .5f * c[0];//the bounceback happens at half-way and is treated as such
+    r[1] = to_center[1] - .5f * c[1];
+    r[2] = to_center[2] - .5f * c[2];
+    loc_v[0] = v[0] - omega[1]*r[2] + omega[2]*r[1];//change of sign since r points to the center.
+    loc_v[1] = v[1] - omega[2]*r[0] + omega[0]*r[2];
+    loc_v[2] = v[2] - omega[0]*r[1] + omega[1]*r[0];
+    vc = loc_v[0]*c[0] + loc_v[1]*c[1] + loc_v[2]*c[2];
+    shift = weight * d_v[to_index].rho[0] * 6.0f*vc;
+    n_curr.vd[inverse*para.number_of_nodes + to_index ] = pop_to_bounce_back + shift;
+    pop_to_bounce_back *= 2;
+    f[0] = (pop_to_bounce_back + shift) * c[0];
+    f[1] = (pop_to_bounce_back + shift) * c[1];
+    f[2] = (pop_to_bounce_back + shift) * c[2];
+    torque[0] -= (to_center[1]*f[2] - to_center[2]*f[1]);//change of sign because r to center
+    torque[1] -= (to_center[2]*f[0] - to_center[0]*f[2]); 
+    torque[2] -= (to_center[0]*f[1] - to_center[1]*f[0]); 
+    boundary_force[0] += f[0];
+    boundary_force[1] += f[1];
+    boundary_force[2] += f[2];
+  }
+
+ 
+#else 
+  for(int component=0; component<LB_COMPONENTS;component++){
+     shift = 2.0f*para.agrid*para.agrid*para.rho[component]*3.0f*weight*para.tau*(v[0]*c[0] + v[1]*c[1] + v[2]*c[2]); 
+     pop_to_bounce_back =  n_curr.vd[(population+component*LBQ)*para.number_of_nodes + index ]; 
+     to_index_x = (x+c[0]+para.dim_x)%para.dim_x; 
+     to_index_y = (y+c[1]+para.dim_y)%para.dim_y; 
+     to_index_z = (z+c[2]+para.dim_z)%para.dim_z; 
+     to_index = to_index_x + para.dim_x*to_index_y + para.dim_x*para.dim_y*to_index_z; 
+     if ( n_curr.boundary[to_index] == 0) 
+     { 
+       boundary_force[0] += (2.0f*pop_to_bounce_back+shift)*c[0]/para.tau/para.tau/para.agrid; 
+       boundary_force[1] += (2.0f*pop_to_bounce_back+shift)*c[1]/para.tau/para.tau/para.agrid; 
+       boundary_force[2] += (2.0f*pop_to_bounce_back+shift)*c[2]/para.tau/para.tau/para.agrid; 
```

Comment:
This is just the expansion of the old [`BOUNCEBACK()`](https://github.com/espressomd/espresso/pull/1201/files/80b4d00c71a44fe04e25db0dbfa7187383d6c394#diff-4e94214b9a5c2058eeed5947f6e317b7L1072) macro.  Torque is probably added elsewhere.